### PR TITLE
build: do not always build seastar as a static library

### PR DESCRIPTION
**Pull request by tchaikov**
* tests/unit: include unit_test.hpp instead of included/unit_test.hpp
* build: do not always build seastar as a static library
* build: build seastar as a shared library in 'debug' or 'dev' mode

Signed-off-by: Kefu Chai <kefu.chai@scylladb.com>

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -740,13 +740,19 @@ target_include_directories (seastar
   PRIVATE
     ${CMAKE_CURRENT_SOURCE_DIR}/src)
 
+option (BUILD_SHARED_LIBS
+  "Build seastar library as shared libraries instead of static"
+  OFF)
```
Comment (avikivity) [inline CMakeLists.txt:745]: Should we default it to ON for dev/debug?

Comment (avikivity) [reply CMakeLists.txt:745]: It probably makes sense for all users, question is what is accepted CMake practice.

Comment (avikivity) [reply CMakeLists.txt:745]: Better idea: let's change ./configure.py to set it to ON for dev/debug. Then it gets exercised during CI and this is useful for users as well.

Comment (tchaikov) [reply CMakeLists.txt:745]: > Should we default it to ON for dev/debug?

no, in theory we should not set an option on behalf of user, and it's OFF by default actually. so this behavior should not be changed at least in CMake.

> Better idea: let's change ./configure.py to set it to ON for dev/debug. Then it gets exercised during CI and this is useful for users as well.

yeah, this makes more sense to me. as applications are more likely to build seastar using CMake, and it inherits parent project's setting if any.

i will create a commit to change `configure.py`.

**Comment by tchaikov** [comment]
changelog

- added a commit to build seastar as a shared library when the build mode is "debug" or "dev".

@avikivity

**Comment by avikivity** [comment]
I inspected the generated .pc files, and they see correct.

**Comment by avikivity** [comment]
Practically everything fails on CI.

**Comment by avikivity** [comment]
Locally it works:

```
$ ldd build/dev/apps/httpd/httpd 
	linux-vdso.so.1 (0x00007ffee7356000)
	libseastar.so => /home/avi/seastar/build/dev/libseastar.so (0x00007f1c1be00000)
	libboost_program_options.so.1.78.0 => /lib64/libboost_program_options.so.1.78.0 (0x00007f1c1c51b000)
	libboost_thread.so.1.78.0 => /lib64/libboost_thread.so.1.78.0 (0x00007f1c1c4fc000)
	libboost_system.so.1.78.0 => /lib64/libboost_system.so.1.78.0 (0x00007f1c1c4f7000)
	libboost_chrono.so.1.78.0 => /lib64/libboost_chrono.so.1.78.0 (0x00007f1c1c4ed000)
	libboost_atomic.so.1.78.0 => /lib64/libboost_atomic.so.1.78.0 (0x00007f1c1c4e1000)
	libcares.so.2 => /lib64/libcares.so.2 (0x00007f1c1c4c7000)
	libcryptopp.so.8 => /lib64/libcryptopp.so.8 (0x00007f1c1b800000)
	libfmt.so.9 => /lib64/libfmt.so.9 (0x00007f1c1c4a6000)
	liblz4.so.1 => /lib64/liblz4.so.1 (0x00007f1c1c483000)
	libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007f1c1b400000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f1c1c3a1000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f1c1c381000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f1c1b223000)
	libgnutls.so.30 => /lib64/libgnutls.so.30 (0x00007f1c1ae00000)
	libsctp.so.1 => /lib64/libsctp.so.1 (0x00007f1c1c37b000)
	libyaml-cpp.so.0.6 => /lib64/libyaml-cpp.so.0.6 (0x00007f1c1bdb0000)
	libhwloc.so.15 => /lib64/libhwloc.so.15 (0x00007f1c1bd53000)
	liburing.so.2 => /lib64/liburing.so.2 (0x00007f1c1bd4c000)
	libnuma.so.1 => /lib64/libnuma.so.1 (0x00007f1c1bd3e000)
	libatomic.so.1 => /lib64/libatomic.so.1 (0x00007f1c1bd34000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f1c1c576000)
	libp11-kit.so.0 => /lib64/libp11-kit.so.0 (0x00007f1c1b6cb000)
	libidn2.so.0 => /lib64/libidn2.so.0 (0x00007f1c1b67e000)
	libunistring.so.2 => /lib64/libunistring.so.2 (0x00007f1c1b06b000)
	libtasn1.so.6 => /lib64/libtasn1.so.6 (0x00007f1c1bd1c000)
	libnettle.so.8 => /lib64/libnettle.so.8 (0x00007f1c1adab000)
	libhogweed.so.6 => /lib64/libhogweed.so.6 (0x00007f1c1b63b000)
	libgmp.so.10 => /lib64/libgmp.so.10 (0x00007f1c1ad06000)
	libffi.so.8 => /lib64/libffi.so.8 (0x00007f1c1bd0e000)
```

**Comment by avikivity** [comment]
Either we introduced or exposed some memory leak in exception_logging_test

**Comment by tchaikov** [comment]
marked as draft before the test failures are fixed.

**Comment by avikivity** [comment]
```console
$ du -sm build/*
3312	build/debug
162	build/dev
11781	build/release
```

**Comment by avikivity** [comment]
But:

```
$ ./build/debug/apps/httpd/httpd 
Segmentation fault (core dumped)
```

**Comment by avikivity** [comment]
```
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x00007fffeea22620 in dl_iterate_phdr (callback=0x7ffff78b8e00 <__asan::FindFirstDSOCallback(dl_phdr_info*, size_t, void*)>, data=0x7fffffffde78) at /home/avi/seastar/src/core/exception_hacks.cc:119
#2  0x00007ffff78b9207 in __asan::AsanCheckDynamicRTPrereqs () at ../../../../libsanitizer/asan/asan_linux.cpp:177
#3  0x00007ffff78c5a93 in __asan::AsanInitInternal () at ../../../../libsanitizer/asan/asan_rtl.cpp:412
#4  0x00007ffff7fcce12 in _dl_init (main_map=0x7ffff7ffe2c0, argc=1, argv=0x7fffffffdf28, env=0x7fffffffdf38) at dl-init.c:102
#5  0x00007ffff7fe38e0 in _dl_start_user () from /lib64/ld-linux-x86-64.so.2
#6  0x0000000000000001 in ?? ()
#7  0x00007fffffffe251 in ?? ()
#8  0x0000000000000000 in ?? ()
```

**Comment by avikivity** [comment]
Probably, we need to disable out dl_iterate_phdr src/core/exception_hacks.cc if building as a shared library.

/cc @gleb-cloudius

**Comment by gleb-cloudius** [comment]
> Probably, we need to disable our dl_iterate_phdr src/core/exception_hacks.cc if building as a shared library.
> 
> /cc @gleb-cloudius

Why? Is it not allowed to provide my own symbols in a dynamic  library? Disabling it will make the dynamic library practically unusable.

The crash is in ASAN though and the function is marked as gnu::no_sanitize_address. Someone ignores our wishes.

**Comment by avikivity** [comment]
> > Probably, we need to disable our dl_iterate_phdr src/core/exception_hacks.cc if building as a shared library.
> > /cc @gleb-cloudius
> 
> Why? Is it not allowed to provide my own symbols in a dynamic library? 

It's problematic if that symbol is used while loading shared libraries. Maybe it's fixable.

> Disabling it will make the dynamic library practically unusable.

It's only for debug/dev mode, we don't expect exception storms on high smp count there.

> 
> The crash is in ASAN though and the function is marked as gnu::no_sanitize_address. Someone ignores our wishes.

Welcome to your life.

**Comment by gleb-cloudius** [comment]
> > > Probably, we need to disable our dl_iterate_phdr src/core/exception_hacks.cc if building as a shared library.
> > > /cc @gleb-cloudius
> > 
> > 
> > Why? Is it not allowed to provide my own symbols in a dynamic library?
> 
> It's problematic if that symbol is used while loading shared libraries. Maybe it's fixable.

I am not sure it is though. Even if we compile seastar as static lib we still link with a lot of dynamic libs and the symbol would have been used.

> 
> > Disabling it will make the dynamic library practically unusable.
> 
> It's only for debug/dev mode, we don't expect exception storms on high smp count there.

It is not a bad idea to use dynamic lib in production as well if possible.

> 
> > The crash is in ASAN though and the function is marked as gnu::no_sanitize_address. Someone ignores our wishes.
> 
> Welcome to your life.

Don't put salt on my wounds.

**Comment by tchaikov** [comment]
could disable `verify_asan_link_order` using 
```console
export ASAN_OPTIONS=verify_asan_link_order=false
```

but i still have 

```
  build/debug/tests/unit/checked_ptr_test
Running 5 test cases...

*** No errors detected
=================================================================
==2883162==ERROR: AddressSanitizer: attempting double-free on 0x6030000024d0 in thread T0:
    #0 0x7fe751eddad8 in operator delete(void*, unsigned long) ../../../../libsanitizer/asan/asan_new_delete.cpp:164
    #1 0x7fe740257946 in __cxa_finalize (/lib64/libc.so.6+0x57946) (BuildId: 7af51ad693d7e1c6a71b7604ded5fb0215e16ec7)
    #2 0x7fe751d958f6  (/lib64/libboost_unit_test_framework.so.1.75.0+0x228f6) (BuildId: a4b5a7a4f8dc9082483b60dd6e70e42dfc6bc875)

0x6030000024d0 is located 0 bytes inside of 18-byte region [0x6030000024d0,0x6030000024e2)
freed by thread T0 here:
    #0 0x7fe751eddad8 in operator delete(void*, unsigned long) ../../../../libsanitizer/asan/asan_new_delete.cpp:164
    #1 0x67e098 in std::__new_allocator<char>::deallocate(char*, unsigned long) /home/kefu/.local/include/c++/13.0.1/bits/new_allocator.h:168
    #2 0x5c3aa4 in std::allocator<char>::deallocate(char*, unsigned long) /home/kefu/.local/include/c++/13.0.1/bits/allocator.h:210
    #3 0x5c3aa4 in std::allocator_traits<std::allocator<char> >::deallocate(std::allocator<char>&, char*, unsigned long) /home/kefu/.local/include/c++/13.0.1/bits/alloc_traits.h:516
    #4 0x5c3aa4 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_destroy(unsigned long) /home/kefu/.local/include/c++/13.0.1/bits/basic_string.h:292
    #5 0x5bdb9e in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_dispose() /home/kefu/.local/include/c++/13.0.1/bits/basic_string.h:286
    #6 0x57459b in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string() /home/kefu/.local/include/c++/13.0.1/bits/basic_string.h:795
    #7 0x7fe740257454 in __run_exit_handlers (/lib64/libc.so.6+0x57454) (BuildId: 7af51ad693d7e1c6a71b7604ded5fb0215e16ec7)

previously allocated by thread T0 here:
    #0 0x7fe751edcbd8 in operator new(unsigned long) ../../../../libsanitizer/asan/asan_new_delete.cpp:95
    #1 0x5bdc07 in std::__new_allocator<char>::allocate(unsigned long, void const*) /home/kefu/.local/include/c++/13.0.1/bits/new_allocator.h:147
    #2 0x56fcd3 in std::allocator<char>::allocate(unsigned long) /home/kefu/.local/include/c++/13.0.1/bits/allocator.h:198
    #3 0x56fcd3 in std::allocator_traits<std::allocator<char> >::allocate(std::allocator<char>&, unsigned long) /home/kefu/.local/include/c++/13.0.1/bits/alloc_traits.h:482
    #4 0x56fcd3 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_create(unsigned long&, unsigned long) /home/kefu/.local/include/c++/13.0.1/bits/basic_string.tcc:155
    #5 0x56ef5c in void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*>(char const*, char const*, std::forward_iterator_tag) /home/kefu/.local/include/c++/13.0.1/bits/basic_string.tcc:225
    #6 0x5268d7 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string<std::allocator<char> >(char const*, std::allocator<char> const&) /home/kefu/.local/include/c++/13.0.1/bits/basic_string.h:642
    #7 0x507288 in __static_initialization_and_destruction_0 /usr/include/boost/test/impl/unit_test_parameters.ipp:97
    #8 0x507e0b in _GLOBAL__sub_I__ZN5boost9unit_test6output22compiler_log_formatter9log_startERSom /home/kefu/dev/seastar/tests/unit/checked_ptr_test.cc:124
    #9 0x7fe74023ffda in __libc_start_main_alias_1 (/lib64/libc.so.6+0x3ffda) (BuildId: 7af51ad693d7e1c6a71b7604ded5fb0215e16ec7)

SUMMARY: AddressSanitizer: double-free ../../../../libsanitizer/asan/asan_new_delete.cpp:164 in operator delete(void*, unsigned long)
==2883162==ABORTING
```

still investigating..

**Comment by avikivity** [comment]
> > > > Probably, we need to disable our dl_iterate_phdr src/core/exception_hacks.cc if building as a shared library.
> > > > /cc @gleb-cloudius
> > > 
> > > 
> > > Why? Is it not allowed to provide my own symbols in a dynamic library?
> > 
> > 
> > It's problematic if that symbol is used while loading shared libraries. Maybe it's fixable.
> 
> I am not sure it is though. Even if we compile seastar as static lib we still link with a lot of dynamic libs and the symbol would have been used.

Yes, but differently. A symbol from a dynamic library overriding a symbol from another dynamic library is different from a  static symbol overriding a symbol from a dynamic library.


> 
> > > Disabling it will make the dynamic library practically unusable.
> > 
> > 
> > It's only for debug/dev mode, we don't expect exception storms on high smp count there.
> 
> It is not a bad idea to use dynamic lib in production as well if possible.

Performance of thread-local variables is terrible, and we use them extensively. It's also bad for link-time optimization.

**Comment by avikivity** [comment]
> could disable `verify_asan_link_order` using
> 
> ```
> export ASAN_OPTIONS=verify_asan_link_order=false
> ```
> 
> but i still have
> 
> ```
>   build/debug/tests/unit/checked_ptr_test
> Running 5 test cases...
> 
> *** No errors detected
> =================================================================
> ==2883162==ERROR: AddressSanitizer: attempting double-free on 0x6030000024d0 in thread T0:
>     #0 0x7fe751eddad8 in operator delete(void*, unsigned long) ../../../../libsanitizer/asan/asan_new_delete.cpp:164
>     #1 0x7fe740257946 in __cxa_finalize (/lib64/libc.so.6+0x57946) (BuildId: 7af51ad693d7e1c6a71b7604ded5fb0215e16ec7)
>     #2 0x7fe751d958f6  (/lib64/libboost_unit_test_framework.so.1.75.0+0x228f6) (BuildId: a4b5a7a4f8dc9082483b60dd6e70e42dfc6bc875)
> 
> 0x6030000024d0 is located 0 bytes inside of 18-byte region [0x6030000024d0,0x6030000024e2)
> freed by thread T0 here:
>     #0 0x7fe751eddad8 in operator delete(void*, unsigned long) ../../../../libsanitizer/asan/asan_new_delete.cpp:164
>     #1 0x67e098 in std::__new_allocator<char>::deallocate(char*, unsigned long) /home/kefu/.local/include/c++/13.0.1/bits/new_allocator.h:168
>     #2 0x5c3aa4 in std::allocator<char>::deallocate(char*, unsigned long) /home/kefu/.local/include/c++/13.0.1/bits/allocator.h:210
>     #3 0x5c3aa4 in std::allocator_traits<std::allocator<char> >::deallocate(std::allocator<char>&, char*, unsigned long) /home/kefu/.local/include/c++/13.0.1/bits/alloc_traits.h:516
>     #4 0x5c3aa4 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_destroy(unsigned long) /home/kefu/.local/include/c++/13.0.1/bits/basic_string.h:292
>     #5 0x5bdb9e in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_dispose() /home/kefu/.local/include/c++/13.0.1/bits/basic_string.h:286
>     #6 0x57459b in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string() /home/kefu/.local/include/c++/13.0.1/bits/basic_string.h:795
>     #7 0x7fe740257454 in __run_exit_handlers (/lib64/libc.so.6+0x57454) (BuildId: 7af51ad693d7e1c6a71b7604ded5fb0215e16ec7)
> 
> previously allocated by thread T0 here:
>     #0 0x7fe751edcbd8 in operator new(unsigned long) ../../../../libsanitizer/asan/asan_new_delete.cpp:95
>     #1 0x5bdc07 in std::__new_allocator<char>::allocate(unsigned long, void const*) /home/kefu/.local/include/c++/13.0.1/bits/new_allocator.h:147
>     #2 0x56fcd3 in std::allocator<char>::allocate(unsigned long) /home/kefu/.local/include/c++/13.0.1/bits/allocator.h:198
>     #3 0x56fcd3 in std::allocator_traits<std::allocator<char> >::allocate(std::allocator<char>&, unsigned long) /home/kefu/.local/include/c++/13.0.1/bits/alloc_traits.h:482
>     #4 0x56fcd3 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_create(unsigned long&, unsigned long) /home/kefu/.local/include/c++/13.0.1/bits/basic_string.tcc:155
>     #5 0x56ef5c in void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*>(char const*, char const*, std::forward_iterator_tag) /home/kefu/.local/include/c++/13.0.1/bits/basic_string.tcc:225
>     #6 0x5268d7 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string<std::allocator<char> >(char const*, std::allocator<char> const&) /home/kefu/.local/include/c++/13.0.1/bits/basic_string.h:642
>     #7 0x507288 in __static_initialization_and_destruction_0 /usr/include/boost/test/impl/unit_test_parameters.ipp:97
>     #8 0x507e0b in _GLOBAL__sub_I__ZN5boost9unit_test6output22compiler_log_formatter9log_startERSom /home/kefu/dev/seastar/tests/unit/checked_ptr_test.cc:124
>     #9 0x7fe74023ffda in __libc_start_main_alias_1 (/lib64/libc.so.6+0x3ffda) (BuildId: 7af51ad693d7e1c6a71b7604ded5fb0215e16ec7)
> 
> SUMMARY: AddressSanitizer: double-free ../../../../libsanitizer/asan/asan_new_delete.cpp:164 in operator delete(void*, unsigned long)
> ==2883162==ABORTING
> ```
> 
> still investigating..

Could be a bug in boost, which we might need to suppress.

**Comment by gleb-cloudius** [comment]
> > > > > Probably, we need to disable our dl_iterate_phdr src/core/exception_hacks.cc if building as a shared library.
> > > > > /cc @gleb-cloudius
> > > > 
> > > > 
> > > > Why? Is it not allowed to provide my own symbols in a dynamic library?
> > > 
> > > 
> > > It's problematic if that symbol is used while loading shared libraries. Maybe it's fixable.
> > 
> > 
> > I am not sure it is though. Even if we compile seastar as static lib we still link with a lot of dynamic libs and the symbol would have been used.
> 
> Yes, but differently. A symbol from a dynamic library overriding a symbol from another dynamic library is different from a static symbol overriding a symbol from a dynamic library.

May be. But I am not sure the symbol is used for dynamic loading. It is used for stack unwinding for sure.

**Comment by tchaikov** [comment]
@avikivity hi Avi, sorry for jumping the gun. should be ready for another round of review.

**Comment by tchaikov** [comment]
changelog

- disable `verify_asan_link_order` in `ASAN_OPTIONS`
- link tests to libboost_unit_test_framework.so
- populate env vars using "env"

**Pull request closed without merging**